gprofiler/utils/node.py

```diff
@@ -0,0 +1,176 @@
+import json
+import os
+import shutil
+import signal
+import stat
+from typing import Dict, List
+
+import psutil
+import requests
+from granulate_utils.linux.ns import get_proc_root_path, get_process_nspid, resolve_proc_root_links, run_in_ns
+from granulate_utils.linux.process import is_musl
+from retry import retry
+from websocket import create_connection
+from websocket._core import WebSocket
+
+from gprofiler.log import get_logger_adapter
+from gprofiler.metadata import get_exe_version
+from gprofiler.utils import add_permission_dir, pgrep_exe, resource_path
+
+logger = get_logger_adapter(__name__)
+
+DSO_GIT_REVISION = "20eb88a"
+
+
+class NodeDebuggerUrlNotFound(Exception):
+    pass
+
+
+class NodeDebuggerUnexpectedResponse(Exception):
+    pass
+
+
+def get_dest_inside_container(musl: bool, node_version: str) -> str:
+    libc = "musl" if musl else "glibc"
+    return os.path.join("/", "tmp", "node_module", DSO_GIT_REVISION, libc, node_version)
+
+
+def _start_debugger(pid: int) -> None:
+    # for windows: in shell node -e "process._debugProcess(PID)"
+    os.kill(pid, signal.SIGUSR1)
+
+
+@retry(NodeDebuggerUrlNotFound, 5, 1)
+def _get_debugger_url() -> str:
+    # when killing process with SIGUSR1 it will open new debugger session on port 9229,
+    # so it will always the same
+    port = 9229
+    debugger_url_response = requests.get(f"http://127.0.0.1:{port}/json/list")
+    if (
+        debugger_url_response.status_code != 200
+        or not debugger_url_response.headers.get("Content-Type")
+        or "application/json" not in debugger_url_response.headers.get("Content-Type")  # type: ignore
+    ):
+        raise NodeDebuggerUrlNotFound(
+            {"status_code": debugger_url_response.status_code, "text": debugger_url_response.text}
+        )
+
+    response_json = debugger_url_response.json()
+    if (
+        not isinstance(response_json, list)
+        or len(response_json) == 0
+        or not isinstance(response_json[0], dict)
+        or "webSocketDebuggerUrl" not in response_json[0]
+    ):
+        raise NodeDebuggerUrlNotFound(response_json)
+
+    return response_json[0]["webSocketDebuggerUrl"]  # type: ignore
+
+
+@retry(NodeDebuggerUnexpectedResponse, 5, 1)
```

Comment:
Why are retries needed?

Reply:
For example, preferably I don't want retries if `_load_dso` fails. What are the retries for?

Reply:
I.e. debugger might not start yet

Reply:
In that case won't we get a connection refused error in the socket level? The current `NodeDebuggerUnexpectedResponse` is an applicative error - an endpoint exists but it responded invalid, I don't understand why trying again would help

Reply:
You are right, I've removed that